cpp/src/arrow/flight/server.cc

```diff
@@ -442,20 +196,69 @@ class GrpcAddCallHeaders : public AddCallHeaders {
   grpc::ServerContext* context_;
 };
 
+class GetDataStream : public internal::TransportDataStream {
+ public:
+  explicit GetDataStream(ServerWriter<pb::FlightData>* writer) : writer_(writer) {}
+
+  Status WriteData(const FlightPayload& payload) override {
+    return internal::WritePayload(payload, writer_);
+  }
+
+ private:
+  ServerWriter<pb::FlightData>* writer_;
+};
+
+class PutDataStream final : public internal::TransportDataStream {
+ public:
+  explicit PutDataStream(grpc::ServerReaderWriter<pb::PutResult, pb::FlightData>* stream)
+      : stream_(stream) {}
+
+  bool ReadData(internal::FlightData* data) override {
+    return internal::ReadPayload(&*stream_, data);
+  }
+  Status WritePutMetadata(const Buffer& metadata) override {
+    pb::PutResult message{};
+    message.set_app_metadata(metadata.data(), metadata.size());
+    if (stream_->Write(message)) {
+      return Status::OK();
+    }
+    return Status::IOError("Unknown error writing metadata.");
+  }
+
+ private:
+  grpc::ServerReaderWriter<pb::PutResult, pb::FlightData>* stream_;
+};
+
+class ExchangeDataStream final : public internal::TransportDataStream {
+ public:
+  explicit ExchangeDataStream(
+      grpc::ServerReaderWriter<pb::FlightData, pb::FlightData>* stream)
+      : stream_(stream) {}
+
+  bool ReadData(internal::FlightData* data) override {
+    return internal::ReadPayload(&*stream_, data);
+  }
+  Status WriteData(const FlightPayload& payload) override {
+    return internal::WritePayload(payload, stream_);
+  }
+
+ private:
+  grpc::ServerReaderWriter<pb::FlightData, pb::FlightData>* stream_;
+};
+
 // This class glues an implementation of FlightServerBase together with the
 // gRPC service definition, so the latter is not exposed in the public API
-class FlightServiceImpl : public FlightService::Service {
+class FlightGrpcServiceImpl : public FlightService::Service {
```

Comment:
Hmm, can you call this something else? It's really too close to `FlightServiceImpl` not to produce any confusion.

Reply:
Renamed to GrpcServiceHandler